session/txn.go

```diff
@@ -227,9 +258,36 @@ func (txn *LazyTxn) changeToInvalid() {
 	txn.stagingHandle = kv.InvalidStagingHandle
 	txn.Transaction = nil
 	txn.txnFuture = nil
-	atomic.StoreUint64(&txn.infoStartTS, 0)
-	atomic.StoreUint64(&txn.EntriesCount, 0)
-	atomic.StoreUint64(&txn.EntriesSize, 0)
+
+	txn.recreateTxnInfo(
+		0,
+		txninfo.TxnRunningNormal,
+		0,
+		0,
+		"",
+		nil)
+}
+
+func (txn *LazyTxn) onStmtStart(currentSQLDigest string) {
+	if len(currentSQLDigest) == 0 {
+		return
+	}
+
+	info := txn.txnInfo.GetSnapInfo()
```

Comment:
Though `GetSnapInfo` is thread safe, accessing `txn.txnInfo` is not, so here's a datarace with `txn.txnInfo = info` in `storeTxnInfo`.

Reply:
onStmtStart  is expected to run within the transaction's thread. The purpose I keeps the `txnInfo` field is to avoid atomic loading every time it's used, and I expect the `txnInfo` is only used within the transaction's thread and the `atomicTxnInfo` is only used for fetching information concurrently. Perhaps this is a bad idea. 😕

Reply:
Ah sorry. I just notice the race reported by CI. It's weird to me that onStmtStart and onStmtEnd is invoked concurrently...

Reply:
The test `TestAmendForUniqueIndex` in pessimistic_test.go is executing two queries concurrently in a single tk...  @cfzjywxk Is this this kind of usage allowed..?

```go
	go func() {
		var err error
		err = tk2.ExecToErr("begin pessimistic")
		if err != nil {
			errCh <- err
			return
		}
		err = tk2.ExecToErr("insert into t values(5, 5)")
		if err != nil {
			errCh <- err
			return
		}
		err = tk2.ExecToErr("delete from t where id = 5")
		if err != nil {
			errCh <- err
			return
		}
		// let commit in tk start.
		errCh <- err
		time.Sleep(time.Millisecond * 100)
		err = tk2.ExecToErr("commit")
		errCh <- err
	}()
	err = <-errCh
	c.Assert(err, Equals, nil)
	tk.MustExec("commit")
	tk2.MustExec("admin check table t")
	err = <-errCh
	c.Assert(err, Equals, nil)
```

Reply:
It's unexpected usage, need to fix.